codegenerator/model/api.go

```diff
@@ -123,76 +141,100 @@ func (auth *Auth) apiCall(payload interface{}, method, route string, result inte
 	}
 	callSummary.HttpRequestBody = string(jsonPayload)
 
-	var ioReader io.Reader = nil
-	if reflect.ValueOf(payload).IsValid() && !reflect.ValueOf(payload).IsNil() {
-		ioReader = bytes.NewReader(jsonPayload)
-	}
-	httpRequest, err := http.NewRequest(method, auth.BaseURL+route, ioReader)
-	if err != nil {
-		callSummary.Error = err
-		return result, callSummary
-	}
-	// only authenticate if client library user wishes to
-	if auth.Authenticate {
-		// not sure if we need to regenerate this with each call, will leave in here for now...
-		credentials := &hawk.Credentials{
-			ID:   auth.ClientId,
-			Key:  auth.AccessToken,
-			Hash: sha256.New,
-		}
-		reqAuth := hawk.NewRequestAuth(httpRequest, credentials, 0).RequestHeader()
-		httpRequest.Header.Set("Authorization", reqAuth)
-	}
-	httpRequest.Header.Set("Content-Type", "application/json")
-	callSummary.HttpRequest = httpRequest
 	httpClient := &http.Client{}
-	// fmt.Println("Request\n=======")
-	// fullRequest, err := httputil.DumpRequestOut(httpRequest, true)
-	// fmt.Println(string(fullRequest))
-	response, err := httpClient.Do(httpRequest)
-	// fmt.Println("Response\n========")
-	// fullResponse, err := httputil.DumpResponse(response, true)
-	// fmt.Println(string(fullResponse))
-	if err != nil {
-		callSummary.Error = err
-		return result, callSummary
+
+	// function to perform http request - we call this using backoff library to
+	// have exponential backoff in case of intermittent failures (e.g. network
+	// blips)
+	httpCall := func() error {
+		var ioReader io.Reader = nil
+		if reflect.ValueOf(payload).IsValid() && !reflect.ValueOf(payload).IsNil() {
+			ioReader = bytes.NewReader(jsonPayload)
+		}
+		httpRequest, err := http.NewRequest(method, auth.BaseURL+route, ioReader)
+		if err != nil {
+			return err
+		}
+		httpRequest.Header.Set("Content-Type", "application/json")
```

Comment:
Only set this if there is a payload.
Also if there is a payload you should set `content-length`, just for good measure... It's nice to have server-side and since these are short JSON objects, there is no reason not to provide it.